certbot/auth_handler.py

```diff
@@ -396,22 +399,20 @@ def _report_no_chall_path(challbs):
 }
 
 
-def _report_failed_challs(failed_achalls):
-    """Notifies the user about failed challenges.
+def _report_failed_authzrs(failed_authzrs, account_key):
+    """Notifies the user about failed authorizations."""
+    problems = {}  # type: Dict[str, List[achallenges.AnnotatedChallenge]]
+    failed_achalls = [challb_to_achall(challb, account_key, authzr.body.identifier.value)
+                      for authzr in failed_authzrs for challb in authzr.body.challenges
+                      if challb.error]
 
-    :param set failed_achalls: A set of failed
-        :class:`certbot.achallenges.AnnotatedChallenge`.
-
-    """
-    problems = collections.defaultdict(list)\
-        # type: DefaultDict[str, List[achallenges.KeyAuthorizationAnnotatedChallenge]]
     for achall in failed_achalls:
-        if achall.error:
-            problems[achall.error.typ].append(achall)
+        problems[achall.error.typ] = problems.get(achall.error.typ, [])
+        problems[achall.error.typ].append(achall)
```

Comment:
I prefer to stay with the bare dict, because it removes a third-party dependency. So nice catch about setdefault. In theory, then I can merge it with the list comprehension above, but I think it would become hard to read.